File: 2025.1/Lab1/src/python/concurrent/sum.py
import os
import sys
import threading

total = 0
sem = threading.Semaphore(1)

def do_sum(path):
    global total
    _sum = 0

    with open(path, 'rb') as f:
        byte = f.read(1)
        while byte:
            _sum += int.from_bytes(byte, byteorder='big', signed=False)
            byte = f.read(1)
        
        with sem:
            total += _sum
            print(path + " : " + str(_sum))

if __name__ == "__main__":
    paths = sys.argv[1:]
    threads = []
    for path in paths:
    #many error could be raised error. we don't care
        new_thread = threading.Thread(target=do_sum, args=(path,))
        threads.append(new_thread)
        new_thread.start()
    
    for thread in threads:
        thread.join()
    
    print(f"Total: {total} bytes")
        

File: 2025.1/Lab1/src/java/concurrent/Sum.java
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.concurrent.Semaphore;

public class Sum {
    private static int total = 0;
    private static Semaphore sem = new Semaphore(1);

    public static void sum(FileInputStream fis, String path) throws IOException {
        
	    int byteRead;
        int sum = 0;

        while ((byteRead = fis.read()) != -1) {
            sum += byteRead;
        }

        try {            
            sem.acquire();
            total += sum;
            System.out.println(path + " : " + sum);
            sem.release();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void sum_p(String path) throws IOException {

        Path f_path = Paths.get(path);

        if (Files.isRegularFile(f_path)) {
            try {
                FileInputStream fis = new FileInputStream(path);
                sum(fis, path);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static void main(String[] args) throws Exception {
        ArrayList<Thread> threads = new ArrayList<Thread>();

        if (args.length < 1) {
            System.err.println("Usage: java Sum filepath1 filepath2 filepathN");
            System.exit(1);
        }

	//many exceptions could be thrown here. we don't care
        for (String path : args) {
            Thread thread = new Thread(new RunnableSum(path), "Thread of path " + path);
            thread.start();
            threads.add(thread);
        }

        for (Thread thread : threads) {
            thread.join();
        }
        
        System.out.println("Total: " + total);
    }
}
